src/core/net/ip6_mpl.cpp

```diff
@@ -85,79 +85,141 @@ void Mpl::InitOption(OptionMpl &aOption, const Address &aAddress)
     }
 }
 
+/*
+ * mSeedSet stores recently received (Seed ID, Sequence) values.
+ * - (Seed ID, Sequence) values are grouped by Seed ID.
+ * - (Seed ID, Sequence) groups are not sorted by Seed ID relative to other groups.
+ * - (Seed ID, Sequence) values within a group are sorted by Sequence.
+ * - All unused entries (marked by 0 lifetime) are grouped at the end.
+ *
+ * Update process:
+ *
+ * - Eviction selection:
+ *   - If there are unused entries, mark the first unused entry for "eviction"
+ *   - Otherwise, pick the first entry of the group that has the most entries.
+ *
+ * - Insert selection:
+ *   - If there exists a group matching the Seed ID, select insert entry based on Sequence ordering.
+ *   - Otherwise, set insert entry equal to evict entry.
+ *
+ * - If evicting a valid entry (lifetime non-zero):
+ *   - Require group size to have >=2 entries.
+ *   - If inserting into existing group, require Sequence to be larger than oldest stored Sequence in group.
+ */
 otError Mpl::UpdateSeedSet(uint16_t aSeedId, uint8_t aSequence)
 {
-    otError       error = OT_ERROR_NONE;
-    MplSeedEntry *entry = NULL;
-    int8_t        diff;
+    otError       error    = OT_ERROR_NONE;
+    MplSeedEntry *insert   = NULL;
+    MplSeedEntry *evict    = mSeedSet;
+    uint8_t       curCount = 0;
+    uint8_t       maxCount = 0;
 
-    for (uint32_t i = 0; i < kNumSeedEntries; i++)
+    for (uint32_t i = 0; i < kNumSeedEntries; i++, curCount++)
     {
         if (mSeedSet[i].GetLifetime() == 0)
         {
-            // Start allocating from the first possible entry to speed up process of searching.
-            if (entry == NULL)
+            // unused entries exist
+
+            if (insert == NULL)
             {
-                entry = &mSeedSet[i];
+                // no existing group, set insert and evict entry to be the same
+                insert = &mSeedSet[i];
             }
-        }
-        else if (mSeedSet[i].GetSeedId() == aSeedId)
-        {
-            entry = &mSeedSet[i];
-            diff  = static_cast<int8_t>(aSequence - entry->GetSequence());
```

Comment:
MinSequence in RFC 7731 does not necessarily map to the minimum sequence received. Instead, MinSequence is intended to indicate the minimum sequence that the device is willing to receive.

This PR technically only implements the MinSequence concept when there are no available entries. This makes sense because entries may be evicted before the lifetime, so the MinSequence rule helps avoid forwarding loops.

When there are entries available, the entry lifetime should be such that forwarding loops are not possible. When the lifetime expires for all entries, that is equivalent to the RFC 7731 Seed Set entry lifetime expiring. At that point, any Sequence value will be accepted again.

Thoughts?

Reply:
Seems reasonable, thanks.